apisix/plugins/elasticsearch-logger.lua

```diff
@@ -0,0 +1,166 @@
+--
+-- Unless required by applicable law or agreed to in writing, software
+-- distributed under the License is distributed on an "AS IS" BASIS,
+-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+-- See the License for the specific language governing permissions and
+-- limitations under the License.
+--
+
+local core            = require("apisix.core")
+local http            = require("resty.http")
+local log_util        = require("apisix.utils.log-util")
+local bp_manager_mod  = require("apisix.utils.batch-processor-manager")
+local plugin          = require("apisix.plugin")
+
+local ngx             = ngx
+local str_format      = core.string.format
+local str_byte        = string.byte
+
+local plugin_name = "elasticsearch-logger"
+local batch_processor_manager = bp_manager_mod.new(plugin_name)
+
+
+local schema = {
+    type = "object",
+    properties = {
+        meta_format = {
+            type = "string",
+            default = "default",
+            enum = {"default", "origin"},
+        },
+        endpoint_addr = core.schema.uri_def,
+        field = {
+            type = "object",
+            properties = {
+                index = { type = "string"},
+                type = { type = "string"}
+            },
+            required = {"index"}
+        },
+        auth = {
+            type = "object",
+            properties = {
+                username = {
+                    type = "string",
+                },
+                password = {
+                    type = "string",
+                },
+            },
+            required = {"username", "password"},
+        },
+        timeout = {
+            type = "integer",
+            minimum = 1,
+            default = 10
+        },
+        ssl_verify = {
+            type = "boolean",
+            default = true
+        }
+    },
+    required = { "endpoint_addr", "field" },
+}
+
+
+local _M = {
+    version = 0.1,
+    priority = 413,
+    name = plugin_name,
+    schema = batch_processor_manager:wrap_schema(schema),
+}
+
+
+function _M.check_schema(conf)
+    return core.schema.check(schema, conf)
+end
+
+
+local function get_logger_entry(conf, ctx)
+    local entry
+    local metadata = plugin.plugin_metadata(plugin_name)
+    core.log.info("metadata: ", core.json.delay_encode(metadata))
+    if metadata and metadata.value.log_format
+    and core.table.nkeys(metadata.value.log_format) > 0
+    then
+        entry = log_util.get_custom_format_log(ctx, metadata.value.log_format)
+        core.log.info("custom log format entry: ", core.json.delay_encode(entry))
+    else
+        entry = log_util.get_full_log(ngx, conf)
+        core.log.info("full log entry: ", core.json.delay_encode(entry))
+    end
+
+    return core.json.encode({
+            create = {
+                _index = conf.field.index,
+                _type = conf.field.type
+            }
+        }) .. "\n" ..
+        core.json.encode(entry) .. "\n"
```

Comment:
I think maybe you can take a try at here, just return JSON here, not JSON string, let me see how elasticsearch handles it.

Reply:
> I think maybe you can take a try at here, just return JSON here, not JSON string, let me see how elasticsearch handles it.

But the `httpc:request_uri` need a string body, if I return `table` here, I still need to do `json.encode` later before `httpc:request_uri`.

Reply:
from: https://github.com/apache/apisix/blob/308bdd967792bb7b34355b98dcf5d6cdeae078f6/apisix/plugins/kafka-logger.lua#L177-L199

the `entry` is just `log_util.get_custom_format_log(ctx, metadata.value.log_format)` or `log_util.get_full_log(ngx, conf)`?

Reply:
> from:
> 
> https://github.com/apache/apisix/blob/308bdd967792bb7b34355b98dcf5d6cdeae078f6/apisix/plugins/kafka-logger.lua#L177-L199
> 
> the `entry` is just `log_util.get_custom_format_log(ctx, metadata.value.log_format)` or `log_util.get_full_log(ngx, conf)`?

But the plugin `kafka-logger` send `string` finally, it do json in this place:
https://github.com/apache/apisix/blob/308bdd967792bb7b34355b98dcf5d6cdeae078f6/apisix/plugins/kafka-logger.lua#L230-L247

I think I must  have failed to understand what you meant in some place.
Do you want to express that return `table` in `log` stage, reassemble it into `json string` in the callback function of `batch_processor_manager` and send it?

Reply:
got it, my mistake.